core/src/main/java/cz/o2/proxima/time/WatermarkEstimator.java

```diff
@@ -15,178 +15,29 @@
-    this.timestampSupplier = Objects.requireNonNull(supplier);
-    this.watermark = new AtomicLong(minWatermark);
-    Preconditions.checkArgument(durationMs > 0, "durationMs must be positive");
-    Preconditions.checkArgument(stepMs > 0, "stepMs must be positive");
-    Preconditions.checkArgument(
-        durationMs / stepMs * stepMs == durationMs, "durationMs must be divisible by stepMs");
-    stepDiffs = new long[(int) (durationMs / stepMs) + 1];
-    for (int i = 0; i < stepDiffs.length; i++) {
-      stepDiffs[i] = 0;
-    }
-    rotatesToInitialize = new AtomicInteger(stepDiffs.length - 1);
-    lastRotate = new AtomicLong(supplier.get() - stepMs);
-  }
+/** Estimates watermark according to incoming stream elements. */
+public interface WatermarkEstimator extends WatermarkSupplier, Serializable {
 
   /**
-   * Accumulate given timestamp at current processing time.
+   * Returns monotonically increasing estimated watermark.
    *
-   * @param stamp the stamp to accumulate
+   * @return the watermark estimate.
    */
-  public void add(long stamp) {
-    rotateIfNeeded();
-    long diff = timestampSupplier.get() - stamp;
-    if (stepDiffs[0] < diff) {
-      stepDiffs[0] = diff;
-    }
-  }
+  long getWatermark();
 
   /**
-   * Retrieve watermark estimate.
+   * Updates the watermark estimate according to the given stream element.
    *
-   * @return the watermark estimate
+   * @param element a stream element.
    */
-  @Override
-  public long getWatermark() {
-    rotateIfNeeded();
-    if (rotatesToInitialize.get() <= 0) {
-      boolean isProcessingBacklog =
-          Arrays.stream(stepDiffs).anyMatch(diff -> diff > allowedTimestampSkew);
-      if (!isProcessingBacklog) {
-        watermark.accumulateAndGet(timestampSupplier.get() - allowedTimestampSkew, Math::max);
-      }
-    }
-    return watermark.get();
-  }
+  default void update(StreamElement element) {};
 
-  private void rotateIfNeeded() {
-    long now = timestampSupplier.get();
-    if (now > lastRotate.get() + stepMs) {
-      rotate(now, (int) ((now - lastRotate.get()) / stepMs));
-    }
-    if (log.isDebugEnabled() && now - lastStatLogged.get() > 10000) {
-      log.debug(
-          "Watermark delay stats: {} with allowedTimestampSkew {}",
-          Arrays.toString(stepDiffs),
-          allowedTimestampSkew);
-      lastStatLogged.set(now);
-    }
-  }
+  /** Signals that streaming source is idle. */
+  default void idle() {}
 
-  private void rotate(long now, int moveCount) {
-    moveCount = Math.min(stepDiffs.length - 1, moveCount);
-    System.arraycopy(stepDiffs, 0, stepDiffs, moveCount, stepDiffs.length - moveCount);
-    if (rotatesToInitialize.get() > 0) {
-      rotatesToInitialize.addAndGet(-moveCount);
-    }
-    for (int i = 0; i < moveCount; i++) {
-      stepDiffs[i] = 0;
-    }
-    lastRotate.set(now);
-  }
+  /** Set min watermark */
+  default void setMinWatermark(long minWatermark) {}
```

Comment:
Javadocs like that have actually no added value. :-)
Can we describe under which conditions is this called? Should it really have default implementation?

Reply:
Right, I improved comment and removed default implementation. It is important concept and rather to force clients to implement it.